server/hedley/modules/custom/hedley_reports/hedley_reports.module

```diff
@@ -3416,6 +3407,969 @@ function hedley_reports_count_acute_illness_signs(array $values) {
+      }
+
+      if (empty($immunisations_by_type[$immunisation_type]['dates'])) {
+        $immunisations_by_type[$immunisation_type]['dates'] = $encounter_dates;
+      }
+      else {
+        $immunisations_by_type[$immunisation_type]['dates'] = array_merge($immunisations_by_type[$immunisation_type]['dates'], $encounter_dates);
+      }
+    }
+  }
+
+  foreach ($immunisations_by_type as $type => $item) {
+    if (empty($item['doses']) || empty($item['dates'])) {
+      continue;
+    }
+
+    $immunisations_by_type[$type]['doses'] = array_unique($item['doses']);
+    sort($immunisations_by_type[$type]['doses']);
+    $immunisations_by_type[$type]['dates'] = array_unique($item['dates']);
+    sort($immunisations_by_type[$type]['dates']);
+    $total_doses = count($immunisations_by_type[$type]['doses']);
+    $total_dates = count($immunisations_by_type[$type]['dates']);
+
+    // Make sure number of doses matches the number of dates.
+    if ($total_doses != $total_dates) {
+      $common_size = min($total_doses, $total_dates);
+      $immunisations_by_type[$type]['doses'] = array_slice($immunisations_by_type[$type]['doses'], 0, $common_size);
+      $immunisations_by_type[$type]['dates'] = array_slice($immunisations_by_type[$type]['dates'], 0, $common_size);
+    }
+
+    foreach ($immunisations_by_type[$type]['doses'] as $index => $dose) {
+      $common_type = hedley_reports_drop_preffix_and_suffix($type, 'well_child_', '_immunisation');
+      $return[$common_type][$dose] = $immunisations_by_type[$type]['dates'][$index];
+    }
+  }
+
+  return $return;
+}
+
+/**
+ * Generates vaccination progress data based on child scoreboard data.
+ *
+ * Retrieves immunization data for a child from the child scoreboard,
+ * considering all encounters up to the current well-child encounter,
+ * and returns structured vaccination progress data.
+ *
+ * @param int $person_id
+ *   The ID of the person (child) whose scoreboard data is being
+ *   processed.
+ * @param DateTime $well_child_encounter_start_date_obj
+ *   The DateTime object representing the start date of the current
+ *   well-child encounter.
+ *
+ * @return array
+ *   A structured array of vaccination doses and dates, organized by
+ *   immunization type, based on child scoreboard encounters.
+ */
+function hedley_reports_well_child_generate_vaccination_progress_data_by_child_scoreboard($person_id, DateTime $well_child_encounter_start_date_obj) {
+  $return = [];
+  if (empty($person_id)) {
+    return $return;
+  }
+
+  $query = new EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'node')
+    ->entityCondition('bundle', 'individual_participant')
+    ->fieldCondition('field_encounter_type', 'value', 'child-scoreboard')
+    ->fieldCondition('field_person', 'target_id', $person_id)
+    ->propertyCondition('status', NODE_PUBLISHED)
+    ->addTag('exclude_deleted');
+
+  $result = $query
+    ->range(0, 1)
```

Comment:
Is it true that one child has always one scoreboard? Or would we need to sort the query?

Reply:
Similar to Nutrition, Child scorecard should have single participant, to which all encounters are pointing.